content/BCPatterns/generic-method-pattern/index.md

```diff
@@ -0,0 +1,327 @@
+
+        if HideDialog then exit(DefaultAnswer);
+        exit(ConfirmManagement.GetResponseOrDefault(ConfirmQst, DefaultAnswer));
+    end;
+
+    local procedure AcknowledgeBlockCustomer(HideDialog: Boolean)
+    var
+        AcknowledgeMsg: Label 'You successfully executed "BlockCustomer"';
+    begin
+        if not GuiAllowed or HideDialog then exit;
+        Message(AcknowledgeMsg);
+    end;
+...
+}
+```
+
+*The Event layer*
+This layer is going to add flexibility to any app that has a dependency on this app.  By default, the pattern always foresees an `OnBefore` and an `OnAfter` event.  
+
+This is the relevant code for the event layer:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        ...
+        OnBeforeBlockCustomer(Cust, Handled);
+        ...
+        OnAfterBlockCustomer(Cust);
+        ...
+    end;
+...    
+    [IntegrationEvent(false, false)]
+    local procedure OnBeforeBlockCustomer(var Cust: Record Customer; var Handled: Boolean);
+    begin
+    end;
+
+    [IntegrationEvent(false, false)]
+    local procedure OnAfterBlockCustomer(var Cust: Record Customer);
+    begin
+    end;
+}
+```
+
+*The method layer*
+The last layer is obviously where the business logic will be written.
+
+The relevant part is:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        ...
+        DoBlockCustomer(Cust, Handled);
+        ...
+    end;
+
+    local procedure DoBlockCustomer(var Cust: Record Customer; Handled: Boolean);
+    begin
+        if Handled then
+            exit;
+
+        Cust.Blocked := Cust.Blocked::All;
+        Cust.Modify(true);
+    end;
+...
+}
+```
+Usually indicated with a "do"-function, the business logic takes place in that procedure.  Obviously, when you have a decent amount of code, it's recommended that you make it readable by applying all the Best Practices in terms of readability in the codeunit.  Though, a few pointers here:
+- keep the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) low
+  - one line (functioncall) after an IF-clause
+  - one line (functioncall) after a repeat
+- readable function calls
+- ...
+
+**Call out to the method from a table**
+
+Now, there is a reason that the one global procedure in the method-codeunit is `internal`.  We shouldn't be calling this procedure directly.  We should be calling it through it's "class" (usually, a table can be seen as a class).  That means, we would have to create a public (not internal) procedure on table-level, which can be used in the business logic.
+
+In our example above, this could be that table extension:
+
+```AL
+tableextension 53100 "Customer Ext BASE" extends Customer
+{
+    procedure BlockCustomer(HideDialog: Boolean)
```

Comment:
Overload function?

```
procedure BlockCustomer()
begin
  BlockCustomer(false);
end;
```

Reply:
Not really the point of the pattern, is it?

Reply:
That's right, but a good habit anyway. And especially for parameters like HideDialog a good idea to implement default behavior that can be overwritten by explicitly providing true as parameter.

Reply:
Seems like a good "guideline" ... ;-)

Reply:
I'll add it .. and I'll add a guideline-issue to be written ;-)